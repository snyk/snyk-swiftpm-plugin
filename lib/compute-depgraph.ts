import { DepGraph, DepGraphBuilder } from '@snyk/dep-graph';
import { execute } from './subprocess';
import * as path from 'path';

export type DepTreeNode = {
  name: string;
  url: string;
  version: string;
  path: string;
  dependencies: DepTreeNode[];
};

export async function computeDepGraph(
  root: string,
  targetFile: string,
  additionalArgs?: string[],
): Promise<DepGraph> {
  const fs = require('fs');
  const os = require('os');

  //create temporary folder
  const tempDir = os.tmpdir();

  //move the .build folder to temporary folder
  fs.renameSync('.build', `${tempDir}/.build`);

  //delete temporary foler
  fs.rmdirSync(tempDir, { recursive: true });

  try {
    const defaultArgs = [
      'package',
      '--package-path',
      path.dirname(targetFile),
      'show-dependencies',
      '--format',
      'json',
    ];

    const args = additionalArgs
      ? defaultArgs.concat(additionalArgs)
      : defaultArgs;
    const result = await execute('swift', args, { cwd: root });
    const depTree: DepTreeNode = JSON.parse(result);
    const depGraph: DepGraph = convertToGraph(depTree);

    return depGraph;
  } catch (e) {
    throw new Error('Unable to generate dependency tree');
  }
}

function traverseTree(
  rootNode: DepTreeNode,
  builder: DepGraphBuilder,
  rootNodeId?: string,
) {
  const childNodes = rootNode.dependencies;

  childNodes?.forEach((node) => {
    const { url, version } = node;
    const name = url
      .replace(/https:\/\//g, '')
      .replace(/http:\/\//g, '')
      .replace(/.git/g, '');
    const parentName =
      rootNodeId ||
      rootNode.url
        .replace(/https:\/\//g, '')
        .replace(/http:\/\//g, '')
        .replace(/.git/g, '');

    const nodeId = `${name}@${version}`;
    const parentNodeId = `${parentName}@${rootNode.version}`;

    builder.addPkgNode({ name, version }, nodeId);
    builder.connectDep(rootNodeId || parentNodeId, nodeId);

    traverseTree(node, builder);
  });
}

function convertToGraph(rootNode: DepTreeNode): DepGraph {
  const { name, version } = rootNode;
  const depGraphBuilder = new DepGraphBuilder(
    { name: 'swift' },
    { name, version },
  );

  traverseTree(rootNode, depGraphBuilder, depGraphBuilder.rootNodeId);

  return depGraphBuilder.build();
}
