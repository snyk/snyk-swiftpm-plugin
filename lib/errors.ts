export class SwiftError extends Error {
    constructor(message: string, swiftErrStack: any) {
        super(message);
        this.stack = swiftErrStack;
    }
}
