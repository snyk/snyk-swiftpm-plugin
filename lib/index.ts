import { computeDepGraph as swiftDepGraph } from './swiftpm/compute-depgraph';
import { computeDepGraph as carthageDepGraph } from './carthage/deps';
import { lookpath } from 'lookpath';
import * as path from 'path';

interface Options {
  debug?: boolean;
  file?: string;
  args?: string[];
}

// we assume that swift considers folders as packages instead of manifest files
function pathToPosix(fpath) {
  const parts = fpath.split(path.sep);
  parts.pop();
  if (parts.length === 0) {
    return './';
  }
  return parts.join(path.posix.sep);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function inspect(
  root: string,
  targetFile: string,
  options?: Options,
) {
  const filename = path.basename(targetFile);
  const supportedPackageManagers = ['Swift Package Manager', 'Carthage'];
  const supportedTargetFiles = {
    'Package.swift': swiftDepGraph,
    'Cartfile.resolved': carthageDepGraph,
  };
  if (!(filename in supportedTargetFiles)) {
    throw new Error(
      `${filename} is not supported by ${supportedPackageManagers.join(' or ')}. ` +
        `Please provide with path to ${Object.keys(supportedTargetFiles).join(' or ')} files.`,
    );
  }
  if (filename === 'Package.swift') {
    const swiftPath = await lookpath('swift');
    if (!swiftPath) {
      throw new Error(
        'The "swift" command is not available on your system. ' +
          'To scan your dependencies in the CLI, you must ensure you have ' +
          'first installed the relevant package manager.',
      );
    }
  }

  const computeDepGraph = supportedTargetFiles[filename];
  const depGraph = await computeDepGraph(root, targetFile, options?.args);
  if (!depGraph) {
    throw new Error(`Failed to scan ${targetFile}`);
  }
  return {
    plugin: {
      name: 'snyk-swiftpm-plugin',
      runtime: 'unknown',
      targetFile: `${pathToPosix(targetFile)}${filename}`,
    },
    dependencyGraph: depGraph,
  };
}
