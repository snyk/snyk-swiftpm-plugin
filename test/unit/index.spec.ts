import { inspect } from '../../lib/index';
import { lookpath } from 'lookpath';
import { computeDepGraph } from '../../lib/compute-depgraph';
import { DepGraphBuilder } from '@snyk/dep-graph';

jest.mock('../../lib/compute-depgraph');
jest.mock('lookpath');
jest.mock('fs');

const depGraph = new DepGraphBuilder(
  { name: 'swift' },
  { name: 'packageName', version: '1.0' },
).build();

describe('inspect', () => {
  it('should return depgraph', async () => {
    const mockedLookpath = jest.mocked(lookpath);
    const mockedComputeDepGraph = jest.mocked(computeDepGraph);
    mockedLookpath.mockResolvedValue('swifty');
    mockedComputeDepGraph.mockResolvedValueOnce(depGraph);

    const result = await inspect(
      `${__dirname}/../fixtures/`,
      'Package.swift',
      {},
    );
    expect(result).toMatchSnapshot();
  });

  it('should reject invalid manifest', async () => {
    const mockedLookpath = jest.mocked(lookpath);
    const mockedComputeDepGraph = jest.mocked(computeDepGraph);
    mockedLookpath.mockResolvedValue('swifty');
    mockedComputeDepGraph.mockResolvedValueOnce(depGraph);
    await expect(() =>
      inspect(`${__dirname}/../fixtures/`, 'manifest.swift', {}),
    ).rejects.toThrowError(
      'manifest.swift is not supported by Swift Package Manager. Please provide with path to Package.swift',
    );
  });

  it('should check for swift binary', async () => {
    const mockedLookpath = jest.mocked(lookpath);
    const mockedComputeDepGraph = jest.mocked(computeDepGraph);
    mockedLookpath.mockResolvedValue(undefined);
    await expect(() =>
      inspect(`${__dirname}/../fixtures/`, 'Package.swift', {}),
    ).rejects.toThrowError(
      'The "swift" command is not available on your system. To scan your dependencies in the CLI, you must ensure you have first installed the relevant package manager.',
    );
  });
});
