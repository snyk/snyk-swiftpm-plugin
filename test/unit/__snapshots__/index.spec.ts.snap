// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect should return depgraph 1`] = `
{
  "dependencyGraph": {
    "graph": {
      "nodes": [
        {
          "deps": [],
          "nodeId": "root-node",
          "pkgId": "packageName@1.0",
        },
      ],
      "rootNodeId": "root-node",
    },
    "pkgManager": {
      "name": "swift",
    },
    "pkgs": [
      {
        "id": "packageName@1.0",
        "info": {
          "name": "packageName",
          "version": "1.0",
        },
      },
    ],
    "schemaVersion": "1.3.0",
  },
  "plugin": {
    "name": "snyk-swiftpm-plugin",
    "runtime": "unknown",
    "targetFile": "./Package.swift",
  },
}
`;
